ARCHITECTURE.md

```diff
@@ -0,0 +1,94 @@
+
+# Bird's Eye View
+[comment]: <> (https://drive.google.com/file/d/1lfEzokkNUJr_JIeSDQfha5a57pokz0qI)
+![Architecture](docs/Architecture.png)
+Lets you define probabilistic graphs or models that can be easily used to compute log probabilities for posterior
+inference or to draw random samples for prior and posterior prediction.
+
+PyMC3 includes a few inference techniques, in particular:
+* Markov chain Monte Carlo
+* Variational Inference
+* Sequential Monte Carlo
+
+It also contains numerous others pieces of functionality such as GraphviZ model visualization tools
+as well as various mathematical helper functions.
+
+The most central pieces functionality of PyMC3 are shown visually below, as well as their
+relation to other major packages. Not all modules are shown, either because
+they are smaller or self explanatory in scope, or they're pending
+deprecation
+
+## Functionality not in PyMC3
+It is easier to start with functionality that is not present in PyMC3 but
+rather deferred to outside libraries. If seeking to understand any
+of the topics below refer to that specific library
+
+### Aesara
+* Gradient computation
+* Random number generation
+* Low level tensor operation definition
+* Low level operation graphs
+
+### ArviZ
+* Plotting e.g. Trace plots, rank plots, posterior plots
+* MCMC sampling diagnostics e.g. Rhat, Effective Sample Size.
+* Model comparison, particularly efficient leave-one-out cross-validation approximation
+* Inference Data structure
+
+
+# Modules
+The codebase of PyMC3 is split among single Python file modules at the root
+level, as well as directories with Python code for logical groups of functionality.
+Admittedly the split between single `.py` module or directory is not defined by a strict
+criteria but tends to occur when single `.py` files would be "too big".
+We will with the modules needed implement "simple MCMC" model shown below
+before detailing the remaining modules, such as Variational Inference, Ordinary Differential Equations,
+or Sequential Monte Carlo.
+
+```python
+with pm.Model() as model:
+  theta = pm.Beta("theta", alpha=1, beta=2)
+  p = pm.Beta("n", p=theta, n=2, observed=[1,2])
+  inf_data = pm.sample()
+
+
+```
+
+## {mod}`pymc3.model`
+Contains primitives related model definition and methods used for evaluation of the model.
+In no particular order they are
+
+* `ContextMeta`: The context manager that enables the `with pm.Model() as model` syntax
+* {class}`~pymc3.Factor`: Defines the methods for the various logprobs for models
+* `ValueGrad` which handles the value and gradient and is the main connection point to Aesara
+* `Deterministic` and `Potential`: Definitions for two pieces of functionality useful in some model definitions
+
+## distributions/
+Contains multiple submodules that define distributions,  as well as logic that aids in distributions usage.
+Currently this module also contains functionality that enables specific model architectures such as BART and Time Series.
```

Comment:
I do not think BART is a "model architecture", instead is a distribution over the space of trees (step-wise functions). I agree is a "weird" distribution because we use data (to define the splits and to get its shape), but this is more about practical convenience. To define a model using BART you need to combine it, at least with one more distribution for the likelihood, so BART is not a model by itself. I think a similar argument applies to time-series.

Reply:
Thanks for comment removed